enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,446 @@
+
+## Summary
+
+OCP customers would like to secure OCP's nodes from external attacks by configuring ingress firewall policies.
+eBPF provides a very flexible filtering mechansim to allow early detection and filtering.
+
+New cluster scoped object `IngressNodeFirewall` will be used to configure node ingress firewall match rules and a per matching rule statistics will be avaliable.
+Admin can provision multiple instances of this object while ensuring certain excuetion order.
+Another cluster scoped object `NodeEndpoint` which will be used to specify the attachment interface(s) for example admin can apply specific node firewall rules on one or more interfaces and another different set of rules on another interface(s) or apply the same rules to all interfaces.
+
+## Motivation
+
+Customers would like to secure OCP cluster from external attacks.
+
+### Goals
+
+- flexible match filters rules for customer to use with a CRD validation to detect and fail invalid configuration.
+- Provide a mechanism for customers to select which nodes via `nodeSelector`.
+- Provide a mechanism to define attachement interfaces for the node ingress firewall objects.  
+- Ingress node firewall is deployed on a fully operational OCP cluster (i.e. cluster comes up with no default ingress firewall policy).
+- fail safe in case of user provisioned invalid rules which can lockup cluster operation.
+
+### Non-Goals
+
+## Proposal
+
+- customers will provision `NodeEndPoint` objects where the nodeFirewall policy will be applied to, if the list is empty the policy will be applied on all interfaces.
+- customers will label nodes where ingress node policy will be deployed on.
+- customers will prepare `IngressNodeFirewall` objects with all required firewall rules and deploy, and via annotation will select the `NodeEndPoint` where the policy will be applied to, if there is no annotation rules will be attached to all interfaces.
+- IngressNodeFW controller will process `IngressNodeFirewall` objects, making sure configuration doesn't contradict with fail safe rules, update eBPF ingress table and attach eBPF kernel hook loaded as `BPF_PROG_TYPE_XDP` to the list of interfaces configured in `NodeEndPoint` object.
+- in order to fail safe and avoid cluster lockup, various pinholes will be open in the firewall to allow basic services.
+
+### Workflow Description
+
+- the following diagram shows how node ingress firewall configuration is handled
+
+![Ingress Firewall configuration processing](./ingress-nodefw-config-flow.png)
+
+- the following diagram shows how eBPF kernel hook is compiled and how the eBPF loader will attached to the interfaces and update the map table with firewall match rules and actions
+
+![eBPF workflow](./eBPF_workflow.png)
+
+- the following diagram shows eBPF how ingress firewall kernel hook will process packets and apply firewall rules
+
+![eBPF workflow](./ebpf-packet-flow.png)
```

Comment:
Am I to understand that you have 2 actions:

1. Drop packets that don't match any rule
2. Drop packets that match a rule where action != allow and update stats

If so, I think you should also collect stats for paclets in `1` also

Reply:
you are right when cluster comes up no fw at all, now once u apply the firewall we will internally prog failSafe rules and if there is no rule for 0.0.0.0/0 we will add one with default deny `action`